batchflow/tests/segmentation_metrics_by_pixels_test.py

```diff
@@ -0,0 +1,312 @@
+"""
+Tests for SegmentationMetricsByPixels class from batchflow.models.metrics.
+Structurally, file consists of two main classes. First one tests metric's output shape, second — its contents.
+Main test functions run for almost all metrics values, declared in metrics_dict (exceptions described below).
+Test data is pre-defined, it's shape and contents were chosen for reasons of balance between visual simplicity
+and test coverage diversity.
+"""
+# pylint: disable=import-error, no-name-in-module
+from itertools import chain
+import numpy as np
+import pytest
+
+from batchflow.models.metrics import SegmentationMetricsByPixels
+
+# Dict {metric_alias : metric_names}
+METRICS_DICT = {'tpr' : ['true_positive_rate', 'sensitivity', 'recall'],
+                'fpr' : ['false_positive_rate', 'fallout'],
+                'fnr' : ['false_negative_rate', 'miss_rate'],
+                'tnr' : ['true_negative_rate', 'specificity'],
+                'prv' : ['prevalence'],
+                # accuracy can't process 'multiclass' parameter and therefore is being tested individually
+                'ppv' : ['positive_predictive_value', 'precision'],
+                'fdr' : ['false_discovery_rate'],
+                'for' : ['false_omission_rate'],
+                'npv' : ['negative_predictive_value'],
+                'plr' : ['positive_likelihood_ratio'],
+                'nlr' : ['negative_likelihood_ratio'],
+                'dor' : ['diagnostics_odds_ratio'],
+                'fos' : ['f1_score', 'dice'],
+                'jac' : ['jaccard', 'iou']}
+# List all metrics' names from METRICS_DICT values.
+METRICS_LIST = list(chain.from_iterable(METRICS_DICT.values()))
+
+BATCH_SIZE = 2
+IMAGE_SIZE = 2
+NUM_CLASSES = 3
+
+# Set targets.
+TARGETS = np.array([0, 1, 2, 2, 0, 0, 1, 1]).reshape(BATCH_SIZE, IMAGE_SIZE, IMAGE_SIZE)
+# Set predictions as 'labels'.
+LABELS = np.array([0, 1, 1, 0, 2, 0, 1, 1]).reshape(BATCH_SIZE, IMAGE_SIZE, IMAGE_SIZE)
+# Onehots are basically like probas, just with all 0 and a single 1.
+PROBA = np.eye(NUM_CLASSES)[LABELS]
+# Logit function gives ±infs on degenerate case of 0s and 1s, but it's okay for sigmoid function.
+LOGITS = np.log(PROBA / (1. - PROBA))
+# First param stands for predictions variable, second — for predictions type, third — for axis with class info.
+PREDICTIONS = [(LABELS, 'labels', None),
+               (PROBA, 'proba', 3),
+               (LOGITS, 'logits', 3)]
```

Comment:
Let's also check different class axes.

Reply:
And thresholds.

Reply:
Axes are now also tested for 'channels_first' data format.